Add support for setting multiple node attributes at the same time

clumens opened the pull request.

kgaillot reviewed (commented):
Yep this looks like a good approach

I'm thinking it will be worthwhile to take a small detour and implement the minimum protocol version tracking first, and enable it for attrd_client_clear_failure(). We could keep a global with the min version, and update it whenever a private attribute is set (in create_attribute()). We could potentially recheck whenever a peer leaves, but that's not necessary.

Then this PR could include multiple-attribute handling in attrd_peer_update() as well, so the changes are at least atomic for IPC queries (they'd be guaranteed to either all be the old values or all be the new values). That would still leave CIB updates non-atomic but that's sufficient for now, since it at least guarantees the CIB writes will be attempted immediately after each other.

clumens commented on lib/common/ipc_attrd.c:
```diff
@@ -94,6 +95,12 @@ dispatch(pcmk_ipc_api_t *api, xmlNode *reply)
     if (pcmk__str_eq((const char *) reply->name, "ack", pcmk__str_none)) {
+        pcmkXmlStr ack_version = xmlGetProp(reply, (pcmkXmlStr) "version");
+
+        if (ack_version && pcmk__numeric_strcasecmp("4", (const char *) ack_version) < 1) {
```
Yep, that is exactly the function I was trying to think of.

clumens commented on lib/common/ipc_attrd.c:
```diff
@@ -492,3 +492,87 @@ pcmk__attrd_api_update(pcmk_ipc_api_t *api, const char *node, const char *name,
+    for (GList *iter = attrs; iter != NULL; iter = iter->next) {
+        pcmk__attrd_query_pair_t *pair = (pcmk__attrd_query_pair_t *) iter->data;
+        const char *target = NULL;
+        xmlNode *child = create_xml_node(request, "op");
```
Fixed.

clumens commented on lib/common/ipc_attrd.c:
```diff
@@ -174,28 +174,40 @@ create_attrd_op(const char *user_name)
     if (rc != pcmk_rc_ok) {
         crm_err("error: Could not connect to attrd: %s", pcmk_rc_str(rc));
-        return ENOTCONN;
+        return NULL;
```
Fixed.

clumens commented on daemons/attrd/pacemaker-attrd.h:
```diff
@@ -139,9 +139,6 @@ typedef struct attribute_value_s {
 extern GHashTable *attributes;
 
-#define attrd_send_ack(client, id, flags) \
-    pcmk__ipc_send_ack((client), (id), (flags), "ack", CRM_EX_INDETERMINATE)
```
Yeah, that looks good.  Fixed.

kgaillot commented on lib/common/ipc_attrd.c:
```diff
@@ -492,3 +492,87 @@ pcmk__attrd_api_update(pcmk_ipc_api_t *api, const char *node, const char *name,
+            rc = pcmk__attrd_api_update(api, pair->node, pair->name, pair->value,
+                                        dampen, set, user_name, options);
+        }
+    }
```
I don't see why not, it's a black box to the caller. We would want to make sure we can't get into an infinite loop (especially if sync), but now that dispatch() returns whether more responses are needed, it should just look like it's waiting on another one.

clumens replied:
Getting back to this...

If I use pcmk__is_daemon to decide whether to serialize requests or not, I was thinking I could just add that into pcmk__attrd_api_update_list so the clients don't have to do that test every time.  The result would more or less be structured like this:

```
int pcmk__attrd_api_update_list(...) {
    request = create_attrd_op(user_name);
    daemon = pcmk__is_daemon();

    add_op_attr(request, options);

    if (api == NULL) {
        create_api(&api);
        created_api = true;
    }

    for (GList *iter = attrs; iter != NULL; iter = iter->next) {
        if (daemon) {
            /* add to multi-op request with populate_update_op */
        } else {
            /* send single op request */
        }
    }

    if (created_api) {
        destroy_api(api);
    }

    return rc;
}
```

Obviously, tons of details are missing there but that's not important.  The important thing is that with this structure, I don't believe we even need the version number on the ACK, do we?  We're making the decision about whether to send multiple or just one big request before ever talking to the server.

kgaillot replied:
Yep that's the approach I had in mind.

I do think it's worthwhile to keep the versioning, it's bound to come in handy sometime.

clumens commented:
Updated with a new version of pcmk__attrd_api_update_list.  Further testing needs to be done before I move this out of draft status, though the "crm_attribute -X" thing I've got in the last commit works as intended.

Also, even if we keep the ACK versioning portion, we could still get rid of the commit that adds private data and the commit that sets whether attrd supports multiple messages.

clumens commented:
Tested the multiple attributes in one message angle by temporarily inverting the tests in [1bd50d1](https://github.com/ClusterLabs/pacemaker/pull/2718/commits/1bd50d18ba149c2f400a3382e39257741fde3fb7) and it looks like it works just fine.

clumens commented:
I've dropped the last patch and renamed and taken this out of draft status.  ctslab testing looks good, and I don't think there's additional testing that can be done until we work this into a daemon somewhere.

kgaillot commented on lib/common/ipc_attrd.c:
```diff
@@ -497,3 +497,96 @@ pcmk__attrd_api_update(pcmk_ipc_api_t *api, const char *node, const char *name,
+     *     The reason we can't use a single message in this case is that we can't
+     *     trust that the server supports it.  Remote nodes could be involved
+     *     here, and there's no guarantee that the version of attrd running on a
+     *     remote node will be recent enough.  We also can't just try sending a
```
attrd doesn't run on remote nodes ... there's no guarantee that a newer client running on a remote is talking to (proxied through) a cluster node with a newer attrd

clumens replied:
Fixed.

kgaillot commented on lib/common/ipc_attrd.c:
```diff
@@ -497,3 +497,96 @@ pcmk__attrd_api_update(pcmk_ipc_api_t *api, const char *node, const char *name,
+     *     and resend as individual messages.
+     *
+     * (2) For messages between daemons, we can be assured that everything will
+     *     have the right version and that a single big message is supported.  So,
```
more specifically, we can be sure that the local attrd will support the new message, and it can send it to the other attrds as one request or split up according to the minimum supported version

not important, but fyi: pacemaker-remoted itself (which is a flavor of pacemaker-execd) would be in the same boat as a remote client, but it doesn't talk to attrd so it doesn't matter

clumens replied:
Fixed.

kgaillot commented on lib/common/ipc_attrd.c:
```diff
@@ -497,3 +497,96 @@ pcmk__attrd_api_update(pcmk_ipc_api_t *api, const char *node, const char *name,
+                pair->node = target;
+            }
+
+            populate_update_op(child, target, pair->name, pair->value, dampen,
```
i think you mean pair->node

clumens replied:
Yep, fixed.

kgaillot commented on daemons/attrd/attrd_commands.c:
```diff
@@ -252,10 +252,29 @@ void
+            attrd_client_update(child);
+        }
+
+        return;
```
We can broadcast the original op instead of splitting it up if all attrds support it. That would arrive at attrd_peer_update()

clumens replied:
Without modifications to attrd_peer_update, I don't see how this is going to work.  That function's going to bail immediately if there is no attribute, and there's no attribute in the top level of the multi-operation XML message.

kgaillot replied:
Yep, attrd_peer_update() is changeable :)

That's the point of tracking the protocol version -- we can know if all peers will support the new syntax or not.

kgaillot reviewed (commented):
I don't think we need the private data anymore, the client side doesn't need to care about support

clumens commented on daemons/attrd/pacemaker-attrd.h:
```diff
@@ -139,8 +139,8 @@ typedef struct attribute_value_s {
-#define attrd_send_ack(client, id, flags) \
-    pcmk__ipc_send_ack((client), (id), (flags), "ack", CRM_EX_INDETERMINATE)
+#define attrd_send_ack(client, id, flags, ver) \
+    pcmk__ipc_send_ack((client), (id), (flags), "ack", (ver), CRM_EX_INDETERMINATE)
```
Fixed.

kgaillot commented on daemons/attrd/attrd_commands.c:
```diff
@@ -955,29 +885,141 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+    attribute_t *a = NULL;
+    const char *attr = crm_element_value(xml, PCMK__XA_ATTR_NAME);
+    const char *value = crm_element_value(xml, PCMK__XA_ATTR_VALUE);
+    gboolean is_force_write = FALSE;
```
this is a case where int/0/1 is appropriate for a boolean since we're passing it to a function that uses int*

clumens replied:
Fixed.

kgaillot commented on daemons/attrd/attrd_commands.c:
```diff
@@ -955,29 +885,141 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+    if (attr == NULL) {
+        crm_warn("Could not update attribute: peer did not specify name");
+        return NULL;
+    }
```
i think this first test is more readable in attrd_peer_update()

clumens replied:
Fixed.

kgaillot commented on daemons/attrd/attrd_commands.c:
```diff
@@ -955,29 +885,141 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+        xml_remove_prop(xml, PCMK__XA_ATTR_NODE_ID);
+        g_hash_table_iter_init(&vIter, a->values);
+        while (g_hash_table_iter_next(&vIter, (gpointer *) & host, NULL)) {
+            attrd_peer_update(peer, xml, host, filter);
```
It should be possible to call update_attr_on_host() here instead

clumens replied:
That change happens in 4a4b9a74a1 but I can move it to this patch if you'd prefer.

kgaillot commented on daemons/attrd/attrd_commands.c:
```diff
@@ -1023,6 +1052,20 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+    if (xml_has_children(xml)) {
+        for (xmlNode *child = first_named_child(xml, XML_ATTR_OP); child != NULL;
+             child = crm_next_same_xml(child)) {
+            attrd_peer_update_one(peer, child, filter);
```
Is it guaranteed that the children always have PCMK__XA_ATTR_NODE_NAME set?

clumens replied:
No, I don't think so.

clumens replied:
Added.

kgaillot reviewed (commented):
Looking good

clumens commented on daemons/attrd/attrd_commands.c:
```diff
@@ -955,29 +885,141 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+    a = populate_attribute(xml, attr);
+    if (attr == NULL) {
+        return;
+    }
```
Fixed.

clumens commented on daemons/attrd/attrd_commands.c:
```diff
@@ -252,10 +252,37 @@ void
+             */
+            for (xmlNode *child = first_named_child(xml, XML_ATTR_OP); child != NULL;
+                 child = crm_next_same_xml(child)) {
+                send_attrd_message(NULL, child);
```
No, just trying to be parallel to what's happening in the top of this block.  But, the comment still refers to attrd_client_update, so I'll change the code.  Fixed.

clumens commented on daemons/attrd/attrd_commands.c:
```diff
@@ -955,29 +885,141 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+
+    // NULL because PCMK__ATTRD_CMD_SYNC_RESPONSE has no PCMK__XA_TASK
+    update_both = pcmk__str_eq(op, PCMK__ATTRD_CMD_UPDATE_BOTH,
+                               pcmk__str_null_matches | pcmk__str_casei);
```
Added a commit on top.

clumens commented on daemons/attrd/attrd_commands.c:
```diff
@@ -914,7 +842,8 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
         crm_notice("Setting %s[%s]: %s -> %s " CRM_XS " from %s",
-                   attr, host, v->current? v->current : "(unset)", value? value : "(unset)", peer->uname);
+                   attr, host, v->current? v->current : "(unset)",
+                   value? value : "(unset)", peer->uname);
```
There were several other places this could be done in this file, and it's easy enough, so I added another patch on top.

kgaillot reviewed (commented):
I think it's just about ready

kgaillot commented on daemons/attrd/attrd_commands.c:
```diff
@@ -1023,6 +1052,25 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+            /* Set the node name on the child message, assuming it isn't already. */
+            if (crm_element_value(xml, PCMK__XA_ATTR_NODE_NAME) == NULL) {
+                crm_xml_add(child, PCMK__XA_ATTR_NODE_NAME, host);
+            }
```
s/xml/child?

clumens replied:
Yes, good catch.  Fixed.
